app/code/community/RicardoMartins/PagSeguro/Model/Abstract.php

```diff
@@ -58,9 +58,21 @@ public function proccessNotificatonResult(SimpleXMLElement $xmlDocument)
             Mage::throwException('Retorno inválido. Referência do pedido não encontrada.');
         }
 
-        /** @var Mage_Sales_Model_Order $order */
-        $order = $notification->getOrder();
-        $payment = $order->getPayment();
+        $payment = $this->getInfoInstance();
+
+        // tries to get order object from payment method, to get
+        // flags setted on the memory but not persisted yet
+        if($payment->getOrder())
+        {
+            $order = $payment->getOrder();
+        }
+        // otherwise, loads the order from notification data and
+        // reloads payment object from order
+        else
+        {
+            $order = $notification->getOrder();
+            $payment = $order->getPayment();
```

Comment:
Esse cara me preocupou um pouco...
$payment = $this->getInfoInstance() devolverá um Mage_Payment_Model_Info.
Mas $order->getPayment() devolverá um Mage_Sales_Model_Order_Payment.

Reply:
A classe Mage_Sales_Model_Order_Payment, assim como a Mage_Sales_Model_Quote_Payment, são extensões da classe Mage_Payment_Model_Info. Esta função é utilizada em cenários diferentes para seus respectivos ambientes - fechamento de pedido para o objeto atrelado ao Quote, e pós venda, objeto atrelado ao pedido.

Quando um objeto de método de pagamento é instanciado, sempre é (deveria ser) passado para ele a referência da classe que o instanciou (Order_Payment ou Quote_Payment), portanto é uma boa prática se referir ao pagamento por meio desta função getInfoInstance. Lidar uma instância de pagamento realizada fora do escopo da criação do método de pagamento pode gerar erros semânticos, portanto só podemos fazer isto em um cenário controlado (oposto do método processNotification da classe Abstract do mód. do PagSeguro, pois ela é chamada em diversos contextos distintos).

Um exemplo de problema: quando a função order da classe Payment_Cc do mód. PagSeguro é executada, a função processNotification é chamada após o retorno da consulta ao webservice da PS. Neste caso, se a função processNotification preenche alguma informação no objeto de pagamento que ela instanciou localmente (por exemplo, um flag no additionalInformation), quando o fluxo de execução retornar à classe Cc e a criação do pedido continuar a ser executada, o objeto de pagamento do Quote - o qual a classe Cc está utilizando - irá passar informações que serão persistidas no objeto de pagamento do pedido, podendo sobrescrever nossas alterações realizadas no processNotification (novamente, additionalInformation é um exemplo fácil de se observar).

Em um cenário ideal, deveríamos chamar o getInfoInstance sem nem mesmo utilizar este if de conferência que eu inseri, mas no módulo a classe Abstract é instanciada diretamente para a chamada desta função processNotification, sem ser carregado o objeto de pagamento associado. Assim, este condicional garante a retrocompatibilidade, mas mantém a concisão semântica pra ocasião da criação do pedido.

Existe um outro erro relatado mais abaixo em que eu confiro se o objeto é da classe Mage_Sales_Model_Order_Payment. Mas note que o erro não consiste em utilizar o getInfoInstance, mas reside na condição do if: a classe correta a utilizar no if é a Mage_Payment_Model_Info.